src/library/scala/collection/mutable/OpenHashMap.scala

```diff
@@ -64,7 +72,22 @@ extends AbstractMap[Key, Value]
   private[this] val actualInitialSize = OpenHashMap.nextPositivePowerOfTwo(initialSize)
 
   private var mask = actualInitialSize - 1
-  private var table : Array[Entry] = new Array[Entry](actualInitialSize)
+
+  /** The state of each of the slots in the hash table.
+    * One of `Empty`, `Deleted` or `Occupied`.
+    */
+  private[this] var slotStates = new Array[Byte](actualInitialSize)
```

Comment:
Did you try to change `OpenEntry` to have a `Value` + `Boolean`, rather than a `Option[Value]`. On my architecture, I think you'd get that for free:

```
$ sbt
sbt> libraryDependencies += "org.openjdk.jol" % "jol-core" % "0.4"

sbt> consoleQuick
...
scala> println(VMSupport.vmDetails())
WARNING: Unable to attach Serviceability Agent: Error: Could not find or load main class org.openjdk.jol.util.HS_SA_Support

WARNING: VM details, e.g. object alignment, reference size, compressed references info will be guessed.

Running 64-bit HotSpot VM.
Using compressed oop with 3-bit shift.
Using compressed klass with 3-bit shift.
Objects are 8 bytes aligned.
Field sizes by type: 4, 1, 1, 2, 2, 4, 4, 8, 8 [bytes]
Array element sizes: 4, 1, 1, 2, 2, 4, 4, 8, 8 [bytes]

scala> ClassLayout.parseClass(Class.forName("scala.collection.mutable.OpenHashMap$OpenEntry")).toPrintable()

"scala.collection.mutable.OpenHashMap.OpenEntry object internals:
 OFFSET  SIZE   TYPE DESCRIPTION                    VALUE
      0    12        (object header)                N/A
     12     4    int OpenEntry.hash                 N/A
     16     4 Object OpenEntry.key                  N/A
     20     4 Option OpenEntry.value                N/A
     24     4 Object OpenEntry.next                 N/A
     28     4        (loss due to the next object alignment)
Instance size: 32 bytes (estimated, the sample instance is not available)
Space losses: 0 bytes internal + 4 bytes external = 4 bytes total
```